### ISSUE TITLE
Travis's curl is too old, causing spurious fetch-test.sh issues

### ISSUE THREAD

**Issue by mnordhoff** [EFForg/https-everywhere#4645]
Spun off from [comments](https://github.com/EFForg/https-everywhere/pull/4619#discussion-diff-59726811) on pull #4619. CC @J0WI.

Travis runs Ubuntu 14.04, which is 1.9 years old and apparently hopelessly out-of-date in Internet-years. ;-) Its (|lib|py)curl version is too old for modern features that we need, causing spurious `fetch-test.sh` failures.

### ECDSA

First and critically, it fails on ECDSA websites, such as an ever-growing number of CloudFlare clients. For example, the `Archive.is` ruleset updated in #4599, which [passed Travis a week ago](https://travis-ci.org/EFForg/https-everywhere/jobs/121946285), would [fail `fetch-test.sh` if proposed today](https://travis-ci.org/mnordhoff/https-everywhere/jobs/123192829):

    INFO Finished comparing http://archive.today/ -> https://archive.today/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://blog.archive.today/ -> https://blog.archive.today/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://www.archive.today/ -> https://www.archive.today/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://archive.is/ -> https://archive.is/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://www.archive.is/ -> https://www.archive.is/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://archive.li/ -> https://archive.li/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://blog.archive.li/ -> https://blog.archive.li/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://www.archive.li/ -> https://www.archive.li/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    ERROR src/chrome/content/rules/Archive.is.xml: Fetch error: http://archive.today/ => https://archive.today/: (35, 'gnutls_handshake() failed: Handshake failed')
    ERROR src/chrome/content/rules/Archive.is.xml: Fetch error: http://blog.archive.today/ => https://blog.archive.today/: (35, 'gnutls_handshake() failed: Handshake failed')
    ERROR src/chrome/content/rules/Archive.is.xml: Fetch error: http://www.archive.today/ => https://www.archive.today/: (35, 'gnutls_handshake() failed: Handshake failed')
    ERROR src/chrome/content/rules/Archive.is.xml: Fetch error: http://archive.li/ => https://archive.li/: (35, 'gnutls_handshake() failed: Handshake failed')
    ERROR src/chrome/content/rules/Archive.is.xml: Fetch error: http://blog.archive.li/ => https://blog.archive.li/: (35, 'gnutls_handshake() failed: Handshake failed')
    ERROR src/chrome/content/rules/Archive.is.xml: Fetch error: http://www.archive.li/ => https://www.archive.li/: (35, 'gnutls_handshake() failed: Handshake failed')
    INFO Finished in 4.51 seconds. Loaded rulesets: 1, URL pairs: 8.
    Test URL test failed.

This would work fine on a current Ubuntu 15.10 system:

    $ ./fetch-test.sh src/chrome/content/rules/Archive.is.xml 
    INFO Finished comparing http://archive.today/ -> https://archive.today/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://blog.archive.today/ -> https://blog.archive.today/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://www.archive.today/ -> https://www.archive.today/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://archive.is/ -> https://archive.is/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://www.archive.is/ -> https://www.archive.is/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://archive.li/ -> https://archive.li/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://blog.archive.li/ -> https://blog.archive.li/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished comparing http://www.archive.li/ -> https://www.archive.li/. Rulefile: src/chrome/content/rules/Archive.is.xml.
    INFO Finished in 10.79 seconds. Loaded rulesets: 1, URL pairs: 8.

### `unrecognized_name`

Second and extremely obscurely, it errors out on receiving an `unrecognized_name` TLS warning alert. This came up in #4619 today, due to https://apt.linode.com/. Older Apache servers will send an `unrecognized_name` warning if the SNI host sent by the client doesn't match `ServerAlias` or `ServerName`:

    $ openssl s_client -connect apt.linode.com:443 -servername apt.linode.com -state
    CONNECTED(00000003)
    SSL_connect:before/connect initialization
    SSL_connect:unknown state
    SSL3 alert read:warning:unrecognized name
    SSL_connect:SSLv3 read server hello A
    ...

The obsolete [RFC 4366 § 3.1](https://tools.ietf.org/html/rfc4366#section-3.1) encouraged this:

       If the server understood the client hello extension but does not
       recognize the server name, it SHOULD send an "unrecognized_name"
       alert (which MAY be fatal).

The current [RFC 6066 § 3](https://tools.ietf.org/html/rfc6066#section-3) does not:

                   If the server understood the ClientHello extension but
       does not recognize the server name, the server SHOULD take one of two
       actions: either abort the handshake by sending a fatal-level
       unrecognized_name(112) alert or continue the handshake.  It is NOT
       RECOMMENDED to send a warning-level unrecognized_name(112) alert,
       because the client's behavior in response to warning-level alerts is
       unpredictable.

For that reason, the alert was [removed in Apache 2.2.30 and 2.4.10 (Apache bug 56241)](https://bz.apache.org/bugzilla/show_bug.cgi?id=56241). Still, as we can see, there is at least one unfortunate 2.2.22 installation in the wild.

Chrome and Firefox both ignore the warning alert. ([SSL Labs reports](https://www.ssllabs.com/ssltest/analyze.html?d=apt.linode.com) "Incorrect SNI alerts".) Ubuntu 14.04's version of curl [fails hard on encountering it](https://travis-ci.org/EFForg/https-everywhere/jobs/123064264). From a [reduced testcase](https://travis-ci.org/mnordhoff/https-everywhere/jobs/123192936):

    INFO Finished comparing http://apt.linode.com/ -> https://apt.linode.com/. Rulefile: src/chrome/content/rules/apt.linode.com.xml.
    ERROR src/chrome/content/rules/apt.linode.com.xml: Fetch error: http://apt.linode.com/ => https://apt.linode.com/: (51, 'gnutls_handshake() warning: The server name sent was not recognized')
    INFO Finished in 0.28 seconds. Loaded rulesets: 1, URL pairs: 1.
    Test URL test failed.

Again, the version of curl in Ubuntu 15.10 ignores it:

    $ ./fetch-test.sh src/chrome/content/rules/apt.linode.com.xml
    INFO Finished comparing http://apt.linode.com/ -> https://apt.linode.com/. Rulefile: src/chrome/content/rules/apt.linode.com.xml.
    INFO Finished in 0.19 seconds. Loaded rulesets: 1, URL pairs: 1.

I'm uncertain, but i believe the curl fix relates to [Debian bug 685402 (2012)](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=685402), and [curl](https://github.com/curl/curl) commits curl/curl@ee3551e45e60856eb0b779aa6cd34d77f16208a5, curl/curl@2045d83dd3f478f7bb8ef86959a82c96235b2bca, curl/curl@41eec4efa2a8c653973b25240c5cda81bb12e26f (all from 2012) and curl/curl@d5aab55b3353bec1d34a2e1434399d23db79b254 (2015-05). I don't know the releases involved.

This Linode rule is very unimportant, and i'm not attached to it, but we will run into more ECDSA problems in the future.

What do?

**Comment by mnordhoff** [comment]
By the way, it's easy to try it from a `python` shell:

    $ sudo aptitude install python-pycurl
    $ python
    import pycurl
    c = pycurl.Curl()
    c.setopt(c.URL, 'https://apt.linode.com/')
    c.perform()

Good: Screenful of HTML.
Bad: `pycurl.error: (51, 'gnutls_handshake() warning: The server name sent was not recognized')` traceback.

To avoid the "screenful of HTML" thing:

    import cStringIO, pycurl
    s = cStringIO.StringIO()
    c = pycurl.Curl()
    c.setopt(c.URL, 'https://apt.linode.com/')
    c.setopt(c.WRITEFUNCTION, s.write)
    c.perform()

(`s.getvalue()` will return the HTML.)

**Comment by J0WI** [comment]
Thanks for debugging this failures!
I think @fuglede is also interested in this issue.

Can we already switch to Ubuntu 16.04 on Travis?

**Comment by mnordhoff** [comment]
I don't think so. They wrote this three weeks ago in travis-ci/travis-ci#5821:

> We do not have immediate plans to update the build environment's release. You can use our Docker feature (https://docs.travis-ci.com/user/trusty-ci-environment) to pull a reasonable image and use that instead.

Another option is whipping up a backported curl package and [configuring Travis to install it](https://docs.travis-ci.com/user/installing-dependencies/). I gave it a shot, but i'm a Debian packaging amateur and got stymied by a trivial dependency issue. I might take another stab at it tomorrow.

**Comment by mnordhoff** [comment]
Got it. Successful Travis run of the ECDSA `Archive.is` ruleset using backported curl and pycurl from Ubuntu 15.10: 

https://travis-ci.org/mnordhoff/https-everywhere/builds/123522486

Branch: https://github.com/mnordhoff/https-everywhere/tree/travis-curl-backport
`.travis.yml` diff: mnordhoff/https-everywhere@9ac2a33a3e00d31a9dda22c0990864c3ddf8304c
PPA: https://launchpad.net/~mnordhoff/+archive/ubuntu/curl-backport

The `pycurl` package is a straight copy of the one in Wily.

The `curl` package was a hassle, but it only took two small `debian/control` changes to fix. I removed RTMP support because dependency hell, and i adjusted a dependency on a newer version of one package because it was only there to work around a bug in a newer version of _another_ package (Debian bugs [784009](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=784009), [787960](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=787960)).

Gist with `control` and diff: https://gist.github.com/mnordhoff/0941ef2219f62762581cdf33c8c5929f

(I don't have a bzr or git repo for the packages, sorry.)

(By the way, the gpg key i used to sign the packages hasn't been signed by anyone. It's listed on my [Launchpad account](https://launchpad.net/~mnordhoff) and [their keyserver](http://keyserver.ubuntu.com:11371/pks/lookup?fingerprint=on&op=index&search=0xEE297CD136C2E2F56CDD7FB51230BD9805F1AC87), but you can't web-of-trust it.)

I'm not saying maintaining backported packages is a good option, or the best option, but it is _an_ option.

**Comment by mnordhoff** [comment]
If my backporting thing is still interesting, i have an Ubuntu 16.04 Xenial version up. (It was just released.)

Branch: https://github.com/mnordhoff/https-everywhere/tree/travis-curl-backport-xenial
`.travis.yml` diff vs. the above: mnordhoff/https-everywhere@49b519a5bb545a1baf8896537157bb74c0f36518
`.travis.yml` diff vs. `master`: https://github.com/EFForg/https-everywhere/compare/master...mnordhoff:travis-curl-backport-xenial?expand=1
PPA (temporary?): https://launchpad.net/~mnordhoff/+archive/ubuntu/curl-backport-xenial
`Archive.is` test Travis: https://travis-ci.org/mnordhoff/https-everywhere/builds/124921813

Packages:

* `curl`. I adjusted a GnuTLS dependency and left RTMP enabled.
* `pycurl`. I diked out its build dependency on the newer package [`python-flaky`](http://packages.ubuntu.com/source/xenial/python-flaky) to avoid backporting half the modern Python infrastructure. flaky was only used by pycurl's test suite — it helps you rerun unreliable tests — and it should be fine without it (but i didn't check).
* `rtmpdump` to provide `librtmp1` for curl. (Also unmodified.)

**Comment by J0WI** [comment]
It seems that this issue breaks most recent submissions.
cc @fuglede @Hainish @jsha

**Comment by Hainish** [comment]
I'm inclined to adopt the backport solution in https://github.com/EFForg/https-everywhere/issues/4645#issuecomment-210770198, but I think we probably want to create a new ppa of our own so we're not dependent on @mnordhoff's.  I'll prevent `travis-ruleset-fetch.sh` from being run until I get a chance to create our own ppa

**Comment by Hainish** [comment]
@J0WI the travis fetch test has been disabled.  Next week I'll have some time to create our own ppa off of @mnordhoff's template and re-enable with backported `curl`

**Comment by fuglede** [comment]
Note also that whatever breakage this might cause will affect <code>https-everywhere-checker</code> as a whole, so if we are catching a large ratio of false negatives here, a lot of rulesets may have potentially been incorrectly disabled on the general database-wide <code>https-everywhere-checker</code> run(s).

**Comment by mnordhoff** [comment]
@Hainish FWIW, Launchpad's most recent package diffs cover everything i changed in the 16.04 packages:

`curl`: https://launchpadlibrarian.net/254163766/curl_7.47.0-1ubuntu2~trusty2_7.47.0-1ubuntu2~trusty3.diff.gz

Ubuntu had renamed `libgnutls28-dev` to `libgnutls-dev` between 14.04 and 16.04 after dropping older versions:

http://packages.ubuntu.com/search?keywords=libgnutls-dev
http://packages.ubuntu.com/search?keywords=libgnutls28-dev

`pycurl`: https://launchpadlibrarian.net/254168601/pycurl_7.43.0-1ubuntu1~trusty1_7.43.0-1ubuntu1~trusty2.diff.gz

I just grepped for "`flaky`" and removed everything. :grinning: It would've been a hassle to backport: [`python-flaky`](http://packages.ubuntu.com/source/xenial/python-flaky) wants things like [`pypy-setuptools`](http://packages.ubuntu.com/xenial/pypy-setuptools) and [`python-genty`](http://packages.ubuntu.com/source/xenial/python-genty), which also wants [`pypy-six`](http://packages.ubuntu.com/xenial/pypy-six). It would've just been a mess, or required some other variety of dependency surgery (for example, to remove the modern PyPy parts).

(Again, i didn't change `rtmpdump`.)

To be honest, i didn't verify my approach was entirely necessary — it might have worked to only backport `pycurl`, skipping `curl`. But 14.04's `curl` is older, and built against an older version of GnuTLS, so backporting everything was *most likely* a good idea.

**Comment by jsha** [comment]
Quick question: Why a PPA vs downloading a built library from curl's website, or downloading source and building it in Travis? We used the latter approach for protocol-compiler in Boulder recently and have been fairly happy with it.

**Comment by Hainish** [comment]
cc @mnordhoff https://github.com/EFForg/https-everywhere/issues/4645#issuecomment-213611894

**Comment by mnordhoff** [comment]
No reason, honestly. Using a PPA was the first approach that came to mind.

Something similar to this would probably work:

    sudo apt-get build-dep curl pycurl
    sudo apt-get install python-pip
    curl -O https://curl.haxx.se/download/curl-7.48.0.tar.bz2
    tar xf curl-7.48.0
    cd curl-7.48.0.tar.bz2
    ./configure --with-gnutls --without-ssl
    make
    # make test
    sudo make install
    sudo pip install pycurl

> downloading a built library from curl's website

Do they make any? For Ubuntu packages, https://curl.haxx.se/download.html just links to packages.ubuntu.com.

**Comment by mnordhoff** [comment]
No reason, honestly. Using a PPA was the first approach that came to mind.

Something similar to this would probably work:

    sudo apt-get build-dep curl pycurl
    sudo apt-get install python-pip
    curl -O https://curl.haxx.se/download/curl-7.48.0.tar.bz2
    tar xf curl-7.48.0.tar.bz2
    cd curl-7.48.0
    ./configure --with-gnutls --without-ssl
    make
    # make test
    sudo make install
    sudo pip install pycurl

Compiling it could add some minutes to the build time, if Travis's VMs aren't very fast. Especially if you run curl's `make test`.

> downloading a built library from curl's website

Do they make any? For Ubuntu packages, https://curl.haxx.se/download.html just links to packages.ubuntu.com.

**Comment by Hainish** [comment]
The problem seems to be not that we're using an outdated version of `curl` and `libcurl`, but that the version of `pycurl` that is installed by default in the travis build system uses `gnutls`, which apparently as @mnordhoff states does not support ECDSA keys.

I found that running the fetch test on https://github.com/EFForg/https-everywhere/blob/master/src/chrome/content/rules/Archive.is.xml on my local instance of Ubuntu 14.04 did not result in an error.  Thinking it curious, I noticed that the errors that we encounter are referencing `gnutls`, but that my own pycurl uses `openssl`.  I had both libraries installed, but it seems that pycurl prefers `openssl` when it is available, unless explicitly told not to.  When travis hits the `pip install -r test/rules/requirements.txt` step, it skips over pycurl because the build environment already has it installed:

```shell
$ pip install -r test/rules/requirements.txt

Requirement already satisfied (use --upgrade to upgrade): pycurl in /usr/lib/python2.7/dist-packages (from -r test/rules/requirements.txt (line 1))
```

This system pycurl in `/usr/lib/python2.7/dist-packages` must have had `gnutls` explicitly preferred.  We can fix this by following up the `pip install -r` commands with a `PYCURL_SSL_LIBRARY=openssl pip install --upgrade --force-reinstall --compile pycurl` command.  This seems to work, using `openssl` instead: https://travis-ci.org/Hainish/https-everywhere/jobs/125960621

I've opened a PR with the fix https://github.com/EFForg/https-everywhere/pull/4683
